pkg/rules/engine/engine.go

```diff
@@ -221,23 +185,47 @@ func (engine *Engine) dispatchEvent(s detect.Signature, event protocol.Event) {
 	engine.signatures[s] <- event
 }
 
-//LoadSignature will store in Engine data structures the given signature and activate its handling business logics.
+//LoadSignature will call the internal signature loading logic and activate its handling business logics.
 // It will return the signature ID as well as error.
 func (engine *Engine) LoadSignature(signature detect.Signature) (string, error) {
+	id, err := engine.loadSignature(signature)
+	if err != nil {
+		return id, err
+	}
+	engine.signaturesMutex.RLock()
+	go signatureStart(signature, engine.signatures[signature], &engine.waitGroup)
+	engine.signaturesMutex.RUnlock()
+
+	return id, nil
+}
+
+//loadSignature handles storing a signature in the Engine data structures
+// It will return the signature ID as well as error.
+func (engine *Engine) loadSignature(signature detect.Signature) (string, error) {
+	metadata, err := signature.GetMetadata()
+	if err != nil {
+		return "", fmt.Errorf("error getting metadata: %v", err)
+	}
 	selectedEvents, err := signature.GetSelectedEvents()
 	if err != nil {
-		return "", fmt.Errorf("failed to store signature: %w", err)
+		return "", fmt.Errorf("error getting selected events for signature %s: %w", metadata.Name, err)
 	}
-	metadata, _ := signature.GetMetadata()
 	// insert in engine.signatures map
-	engine.signaturesMutex.Lock()
-	defer engine.signaturesMutex.Unlock()
+	engine.signaturesMutex.RLock()
```

Comment:
Not in scope of this PR, but I'm just wondering why haven't used sync.Map for signatures index. It would probably make this code much easier to reason about.

Reply:
Sync.Map uses `any` as it's types though right? I think it would make code cleaner but there might be overhead in accessing the signatures that way, maybe we should attempt this with generics, if and when that's supported in `sync.Map` or implement our own sync-safe `SignatureMap`, WDYT?

Reply:
Get started with SitnagureMap seems to me like a good idea. To see if it improves the readability of code. For Go 1.18 with generics I'm not familiar with performance implications, which might be similar to sync.Map.

Reply:
Alright, can we do it in another PR though? I'll open a relevant issue if so (and maybe we should add a "refactor" label to issues?)

Reply:
Like I said in my initial commit this is out of scope of this PR.